### PR TITLE
Added a way to check if players are abusing it to cheat death

### PR DESCRIPTION
**Pull request by Ghoulboy78**
Basically, it won't work if you are in or above lava or fire, and if you are falling through the air. Please tell me if I should add other things.

### CODE REVIEW

File: spectator/data/spectator/functions/spectator.mcfunction [hunk 1]
```diff
@@ -1,8 +1,23 @@
...
\ No newline at end of file
+#Conditions for cheating (block is fire/lava, or falling)
+execute if block ~ ~ ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~ ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~1 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~1 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~2 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~2 ~ minecraft:fire run scoreboard players set @s valid 0
```
Comment (Pablete1234) [inline spectator/data/spectator/functions/spectator.mcfunction]: Could just use the Fire NBT tag instead. It should be at -20 if you're not in lava nor fire. I don't think it's worth checking so many blocks

File: spectator/data/spectator/functions/spectator.mcfunction [hunk 2]
```diff
@@ -1,8 +1,23 @@
...
+execute if block ~ ~-1 ~ minecraft:air unless data entity @s {FallDistance:0f} run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:cactus run scoreboard players set @s valid 0
```
Comment (Pablete1234) [inline spectator/data/spectator/functions/spectator.mcfunction]: I think this one isn't really needed. Being ontop of a cactus will rarely kill you and it's easier to step off one than to trigger into spectator

File: spectator/data/spectator/functions/spectator.mcfunction [hunk 3]
```diff
@@ -1,8 +1,23 @@
...
+execute if data entity @s {Air:0s} run scoreboard players set @s valid 0
```
Comment (Pablete1234) [inline spectator/data/spectator/functions/spectator.mcfunction]: I think it's better to enforce air=300. If you're under water you shouldn't be triggering spectator. Recover your air  first.

File: spectator/data/spectator/functions/spectator.mcfunction [hunk 4]
```diff
@@ -1,8 +1,23 @@
...
+#Conditions for cheating (block is fire/lava, or falling)
+execute if block ~ ~ ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~ ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~1 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~1 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~2 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~2 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:air unless data entity @s {FallDistance:0f} run scoreboard players set @s valid 0
```
Comment (Pablete1234) [inline spectator/data/spectator/functions/spectator.mcfunction]: If you're trapped in a cobweb (potentially going to die from fall or lava later), FallDistance is 0. Consider using OnGround instead.
You'll need to check for OnGround OR current block (or block under you) is water.

**Review by Pablete1234 — CHANGES_REQUESTED** [review]
Checking for danger or inminent death is generally a very hard thing, there are so many things that could put you in danger that it's impossible to check for all of them, but we can definetly check for falling, fire and drawning.

You'll also need to inform the player that they're in too much danger to get into spectator, otherwise they'll think the datapack is broken if they try to use it under water or something similar

**Comment by Ghoulboy78** [comment]
It think I've done all those changes @Pablete1234 , please review and tell me if I've done something wrong.
Also, should I make it so you can't do it if you are near mobs? Cos otherwise you could use it to escape mob fights.

**Comment by Pablete1234** [comment]
Looking for entity data is going to get annoying & give potential falsle positives, they could be near trapped mobs that are irrelevant

File: spectator/data/spectator/functions/spectator.mcfunction [hunk 5]
```diff
@@ -1,8 +1,16 @@
...
 execute store result score @s spec_z run data get entity @s Pos[2] 8
 execute store result score @s spec_rx run data get entity @s Rotation[0] 64
 execute store result score @s spec_ry run data get entity @s Rotation[1] 64
 execute store result score @s spec_dim run data get entity @s Dimension
 
-gamemode spectator @s
\ No newline at end of file
+#Conditions for cheating (block is fire/lava, or falling)
+execute unless data entity @s {Air:300s} run scoreboard players set @s valid 0
+execute unless data entity @s {OnGround:1b} run scoreboard players set @s valid 0
```
Comment (Pablete1234) [inline spectator/data/spectator/functions/spectator.mcfunction:11]: This won't let you go into spectator if floating on water. If you passed the Air:300 check this should allow for water

Comment (Ghoulboy78) [reply spectator/data/spectator/functions/spectator.mcfunction:11]: and exactly how do you float on water?

**Comment by Ghoulboy78** [comment]
Also, about the mob thing, that's why I asked, cos I thought about whether it could lead to unwanted results. Also,should we check if you're flying on an elytra?

File: spectator/data/spectator/functions/spectator.mcfunction (hunk 5 above)
Comment (Pablete1234) [reply spectator/data/spectator/functions/spectator.mcfunction:11]: Pressing spacebar

Comment (Ghoulboy78) [reply spectator/data/spectator/functions/spectator.mcfunction:11]: but how do you even enter the command while having spacebar held down? cos when you press T to open chat, you begin to sink

Comment (Pablete1234) [reply spectator/data/spectator/functions/spectator.mcfunction:11]: You sink slower than you can type, and the command is ran up to 1s later than you entering it. You have the time it takes to sink from feet to head to get into spec, any later will get you denied for Air time (drowning), but the point is being on water should pass the "on ground" check

Comment (Ghoulboy78) [reply spectator/data/spectator/functions/spectator.mcfunction:11]: @Pablete1234 Please view commit 0ae9ad4 on 12th march to see if that's what you mean

**Review by Pablete1234 — APPROVED** [review]

**Pull request merged**